src/drivers/scsi/scsi.c

```diff
@@ -22,81 +22,87 @@ static inline struct usb_mass *scsi2mass(struct scsi_dev *dev) {
 	return member_cast_out(dev, struct usb_mass, scsi_dev);
 }
 
-static int scsi_cmd(struct scsi_dev *sdev, void *cmd, size_t cmd_len,
-		void *data, size_t data_len) {
+int scsi_do_cmd(struct scsi_dev *sdev, struct scsi_cmd *scmd) {
 	struct usb_mass *mass = scsi2mass(sdev);
-	struct scsi_cmd *scmd = cmd;
 	enum usb_direction usb_dir;
+	uint8_t scmd_opcode;
 
 	if (!sdev->attached) {
 		return -ENODEV;
 	}
-	if (scmd->scmd_opcode == SCSI_CMD_OPCODE_WRITE10) {
+	assert(scmd->scmd_len <= SCSI_CMD_MAX_LEN);
+	assert(scmd->scmd_buf);
+
+	scmd_opcode = scmd->scmd_buf[0];
+	if (scmd_opcode == SCSI_CMD_OPCODE_WRITE10) {
 		usb_dir = USB_DIRECTION_OUT;
 	} else {
 		usb_dir = USB_DIRECTION_IN;
 	}
+	log_debug("opcode 0x%x, usb_dir(%d), len %d",
+		scmd_opcode, usb_dir, scmd->scmd_len);
 
-	log_debug("opc (0x%x) cmd_len(%d), usb_dir(%d), data_len(%d)",
-		    scmd->scmd_opcode, cmd_len, usb_dir, data_len);
-	
-	return usb_ms_transfer(mass->usb_dev, cmd, cmd_len, usb_dir, data, data_len);
+	return usb_ms_transfer(mass->usb_dev, scmd->scmd_buf, scmd->scmd_len,
+		       usb_dir, scmd->sdata_buf, scmd->sdata_len);
 }
 
-int scsi_do_cmd(struct scsi_dev *dev, struct scsi_cmd *cmd) {
-	uint8_t scmd[SCSI_CMD_MAX_LEN];
-
-	log_debug("opcode 0x%x, len %d", cmd->scmd_opcode, cmd->scmd_len);
-
-	assert(cmd->scmd_len <= SCSI_CMD_MAX_LEN);
-	memset(scmd + 1, 0, cmd->scmd_len - 1);
-	scmd[0] = cmd->scmd_opcode;
-
-	if (cmd->scmd_fixup) {
-		cmd->scmd_fixup(scmd, dev, cmd);
-	}
+static int scsi_send_inquiry(struct scsi_dev *dev) {
+	struct scsi_cmd cmd;
+	struct scsi_cmd_inquiry cmd_inquiry;
+	struct scsi_data_inquiry data;
 
-	return scsi_cmd(dev, scmd, cmd->scmd_len, cmd->scmd_obuf, cmd->scmd_olen);
-}
+	memset(&cmd_inquiry, 0, sizeof cmd_inquiry);
+	cmd_inquiry.sinq_opcode = SCSI_CMD_OPCODE_INQUIRY;
+	cmd_inquiry.sinq_alloc_length = htobe16(cmd.sdata_len);
 
-static int scsi_send_inquiry(struct scsi_dev *dev) {
-	struct scsi_cmd cmd = scsi_cmd_template_inquiry;
-	struct scsi_data_inquiry *data;
+	cmd.scmd_buf    = (uint8_t *) &cmd_inquiry;
+	cmd.scmd_len    = sizeof cmd_inquiry;
+	cmd.sdata_buf   = (uint8_t *) &data;
+	cmd.sdata_len   = sizeof data;
 
-	cmd.scmd_obuf = dev->scsi_data_scratchpad;
-	cmd.scmd_olen = USB_SCSI_SCRATCHPAD_LEN;
 	if (scsi_do_cmd(dev, &cmd) < 0) {
 		return -1;
 	}
-
-	data = (struct scsi_data_inquiry *) dev->scsi_data_scratchpad;
-	if ((data->dinq_devtype & SCSI_INQIRY_DEVTYPE_MASK)
+	if ((data.dinq_devtype & SCSI_INQIRY_DEVTYPE_MASK)
 			!= SCSI_INQIRY_DEVTYPE_BLK) {
 		return -1;
 	}
 	return 0;
 }
 
 static int scsi_send_test_unit(struct scsi_dev *dev) {
-	struct scsi_cmd cmd = scsi_cmd_template_test_unit;
+	struct scsi_cmd cmd;
```

Comment:
`cmd.sdata_buf` is not set here, so it may contain garbage from stack.

As I figured out, this address may end up being passed to dcache_flush here:
```
ehci_hcd.c:680 dcache_flush(req->buf, req->len);
```

It this address will be unaligned, `dcache_flush()` will try to align and flush it even if `size == 0`
```
void dcache_flush(const void *p, size_t size) {
...
	start &= ~(line_size - 1);
	while (start < end) {
		asm volatile ("mcr p15, 0, %0, c7, c14, 1" : : "r" (start));
		start += line_size;
	}
}
```

This may lead to unwanted flushing of random memory fragments.

Reply:
I have similar concerns for `struct scsi_data_sense data;` & `struct scsi_data_cap10 data;` not being zeroed, but I can't make up example when this may fail for currently implemented drivers.

Reply:
Ok, thank you, I fill these structures with zeroes.